src/jdk.jdwp.agent/share/native/libjdwp/classTrack.c

```diff
@@ -213,13 +232,15 @@ cleanDeleted(void *signatureVoid, void *arg)
 void
 classTrack_reset(void)
 {
+    struct bag* to_delete = NULL;
     debugMonitorEnter(classTrackLock);
 
     if (deletedSignatures != NULL) {
         bagEnumerateOver(deletedSignatures, cleanDeleted, NULL);
-        bagDestroyBag(deletedSignatures);
+        to_delete = deletedSignatures;
```

Comment:
`cleanDeleted` also calls `jvmtiDeallocate`, so either both need to be outside the lock or neither.

It is really unclear to me how many threads can be involved here and which functions can be called when.

Reply:
Right. Moved `bagEnumerateOver()` outside of lock as well.